### ISSUE TITLE
New Rule: no-top-level-await

### ISSUE THREAD

**Issue by kirkwaiblinger** [eslint/eslint#19481]
### Rule details

prohibit TLA

### Related ECMAScript feature

[require(esm) unflagged in node 23](https://nodejs.org/docs/latest/api/modules.html#loading-ecmascript-modules-using-require)

### What type of rule is this?

Warns about a potential problem

### Example code

```js
// bad

await null;

const inAnExpression = Math.random() > 0.5 ? await null : undefined;

for await (const x of asyncGenerator()) {
}

await using foo = asyncDisposable();

// good

async function f() {
  await null;
}

async function* g() {
  for await (const x of foo()) {
  }
}

const arrow = async () => {
    await using x = y;
}

const functionExpression = async function () { 
    await null;
}
```

### Why should this rule be in the core instead of a plugin?

This was previously proposed in https://github.com/eslint/eslint/issues/15339. However, I'd like to offer the following thoughts on why to reconsider it now.

1. It is quite difficult to do with no-restricted-syntax. My attempt is `"AwaitExpression:not(:matches(FunctionExpression AwaitExpression, ArrowFunctionExpression AwaitExpression, FunctionDeclaration AwaitExpression)), ForOfStatement[await=true]:not(:matches(FunctionExpression ForOfStatement[await=true], ArrowFunctionExpression ForOfStatement[await=true], FunctionDeclaration ForOfStatement[await=true])), VariableDeclaration[kind=\"await using\"]:not(:matches(FunctionExpression VariableDeclaration[kind=\"await using\"], ArrowFunctionExpression VariableDeclaration[kind=\"await using\"], FunctionDeclaration VariableDeclaration[kind=\"await using\"]))"`. Maybe this can be made better, but it's regardless too much to ask a user to implement, I think.

2. This is newly important with `require(esm)` becoming a widely available feature. People who have a CJS codebase and would like to use `require(esm)` to incrementally transition to ESM may encounter difficult-to-debug runtime errors ([that not even TS will presently help catch](https://github.com/microsoft/TypeScript/pull/60761)) if they inadvertently use TLA in the ESM portion of the codebase.  Furthermore, package authors may begin authoring packages in ESM only and tell downstream CJS users to just `require()` them if they are still using CJS. In this case, the package author would benefit from banning TLA, so as not to poison the downstream dependency graph.

Of course this rule would need to have a caveat that TLA is perfectly valid in some scenarios (running node.js CLI scripts being an obvious example); this is a tool only to be used in the scenarios in which you have a reason to want to ban TLA.

### Participation

- [x] I am willing to submit a pull request to implement this rule.

### Additional comments

initially came up in https://github.com/typescript-eslint/typescript-eslint/issues/10920 but we realized the no-restricted-syntax version was incomplete, and this is much more widely applicable than TS.

**Comment by ota-meshi** [comment]
If you want to check a top-level-await early, the plugin allows you to check for it.

https://eslint-community.github.io/eslint-plugin-es-x/rules/no-top-level-await.html

**Comment by JoshuaKGoldberg** [comment]
Thinking: perhaps TLA should only be banned in a file with >=1 `export`? As stated in the OP, using TLA in a "leaf node" file that imports modules is fine practice for executables and the like. It's only when the TLA slows down your module imports tree that's the problem AFAIK.

**Comment by kirkwaiblinger** [comment]
That sounds like a plausible idea at first glance but I'm wondering if the existence of side-effect imports throws a fatal wrench into that heuristic 🤔

**Comment by kirkwaiblinger** [comment]
But yeah, this rule would be totally appropriate for some files and totally inappropriate for others. At a minimum the docs would need to make the applicability very clear, even if it's entirely the user's problem to make the call.

**Comment by mdjermanovic** [comment]
> 2\. This is newly important with `require(esm)` becoming a widely available feature. People who have a CJS codebase and would like to use `require(esm)` to incrementally transition to ESM may encounter difficult-to-debug runtime errors ([that not even TS will presently help catch](https://github.com/microsoft/TypeScript/pull/60761)) if they inadvertently use TLA in the ESM portion of the codebase.  Furthermore, package authors may begin authoring packages in ESM only and tell downstream CJS users to just `require()` them if they are still using CJS. In this case, the package author would benefit from banning TLA, so as not to poison the downstream dependency graph.

This is a Node.js-specific problem and therefore I think the rule would be more appropriate for `eslint-plugin-n` than the core. Also, as mentioned by @ota-meshi, this rule already exists in `eslint-plugin-es-x`.

So, I'm not in favor of adding this rule to core ESLint. @eslint/eslint-team thoughts?

**Comment by JoshuaKGoldberg** [comment]
> This is a Node.js-specific problem

I think banning exported TLAs is relevant in all runtimes. TLA just has become more visible now that Node has enabled `require(esm)`. Example from Bun (https://bun.sh/docs/runtime/modules#top-level-await): TLA in a module blocks being able to `require()` it.

This is also an issue in browser code. Putting a TLA in an imported module has similar issues with slowing down running the module dependency graph. Exported TLAs aren't supported well by all tools (e.g. https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/729) but in general do work.

**Comment by nzakas** [comment]
In general, we don't add new rules that just disallow specific syntax. These are simple enough for people to write on their own (as evidenced by there already existing a plugin rule for this purpose).

I agree with @mdjermanovic that I think this is still primarily a server-side runtime issue and I don't think there's a strong case to ban top-level await in most other contexts, including the browser. (Blocking behavior occurs all over the place in browsers, this isn't unique to TLA and something to manage rather than all-out ban.)

I think adding a rule to `eslint-plugin-n` and enabling it as part of the recommended config is the better approach.

**Comment by nzakas** [comment]
It looks like there's not enough support on the TSC to move forward with this, so closing.

If we want to add a new rule to `eslint-plugin-n`, someone can feel free to open a PR there.

**Comment by levrik** [comment]
@JoshuaKGoldberg 

> Exported TLAs aren't supported well by all tools (e.g. https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/729) but in general do work.

This has been fixed in their latest release.
